### DISCUSSION TITLE
Forms (Windows) constructors using defines

### DISCUSSION THREAD

**Discussion by rossanoparis** [KeyWorksRW/wxUiEditor#841] [Ideas]
Using the generated code for a Form such as Dialog, Panel, ecc ... The resulting code is a bit unfriendly in case of multiple instances with different constructor parameters.

For instance, suppose to instance three or even more different dialogues, within an extended code, with different window name only.
The programmer is forced to rewrite all parameters such as position, style, size.
If someone redefines the dialog size, after its first design, other project contributors have to remember where to change it, along the entire project code.

**Actually this is the code I should write**
```
CMyDialogBase m_Instance1(parent, ID_MYDIALOG, _("Title 1"), wxPoint(500,500), wxSize(800, 600), wxCLOSE_BOX|wxRESIZE_BORDER, "MyName1");
...
CMyDialogBase m_Instance2(parent, ID_MYDIALOG, _("Title 2"), wxPoint(500,500), wxSize(800, 600), wxCLOSE_BOX|wxRESIZE_BORDER, "MyName2");
...
CMyDialogBase m_Instance1(parent, ID_MYDIALOG, _("Title 3"), wxPoint(500,500), wxSize(800, 600), wxCLOSE_BOX|wxRESIZE_BORDER, "MyName3");
```

**#define as constructors parameters**
Below, my suggestion

![image](https://user-images.githubusercontent.com/33904565/204313778-e91e552e-95be-4715-ae7f-fc69d86ee4ff.png)

**Resulting header code**
```
#include <wx/dialog.h>
#include <wx/gdicmn.h>
#include <wx/intl.h>

#define CMYDIALOGBASE_ID wxID_HIGHEST + 1
#define CMYDIALOGBASE_TITLE _("MyDialogBase")
#define CMYDIALOGBASE_STYLE wxCLOSE_BOX|wxRESIZE_BORDER
#define CMYDIALOGBASE_POSITION wxPoint(500, 500)
#define CMYDIALOGBASE_SIZE wxSize(800, 600)
#define CMYDIALOGBASE_NAME _("MyDialogName")

class CMyDialogBase : public wxDialog
{
public:
    CMyDialogBase() {}
    CMyDialogBase(wxWindow *parent, wxWindowID id = ID_MYDIALOG, const wxString& title = CMYDIALOGBASE_TITLE,
        const wxPoint& pos = CMYDIALOGBASE_POSITION, const wxSize& size = CMYDIALOGBASE_SIZE,
        long style = CMYDIALOGBASE_STYLE, const wxString &name = CMYDIALOGBASE_NAME)
    {
        Create(parent, id, title, pos, size, style, name);
    }

    bool Create(wxWindow *parent, wxWindowID id = ID_MYDIALOG, const wxString& title = CMYDIALOGBASE_TITLE,
        const wxPoint& pos = CMYDIALOGBASE_POSITION, const wxSize& size = CMYDIALOGBASE_SIZE,
        long style = CMYDIALOGBASE_STYLE, const wxString &name = CMYDIALOGBASE_NAME);

    enum
    {
        ID_MYDIALOG = CMYDIALOGBASE_ID
    };

protected:
};
```

**Instances using defines within an extended project**
```
CMyDialogBase m_Instance1(parent, ID_MYDIALOG, _("Title 1"), CMYDIALOGBASE_POSITION, CMYDIALOGBASE_SIZE, CMYDIALOGBASE_STYLE, "MyName1");
...
CMyDialogBase m_Instance2(parent, ID_MYDIALOG, _("Title 2"), CMYDIALOGBASE_POSITION, CMYDIALOGBASE_SIZE, CMYDIALOGBASE_STYLE, "MyName2");
...
CMyDialogBase m_Instance1(parent, ID_MYDIALOG, _("Title 3"), CMYDIALOGBASE_POSITION, CMYDIALOGBASE_SIZE, CMYDIALOGBASE_STYLE, "MyName3");
```

**Comment by rossanoparis** [comment]
Thank you @KeyWorksRW 

I get your point, yet apart for IDs which is not a big deal, the rest seems a "work around" that could be acceptable, eventhough the way I wanted to expose, could solve every programmer needs.
To garantee a versatile code generated by the RAD tool (that is the point), you could generate such defines and "simply" keep them aligned from their window properties, then is up to the user how to use them.

**Comment by rossanoparis** [comment]
@KeyWorksRW, it's perfect, that's exactly what I meant. 
Indeed, the way you suggest to achive it, is much better than using defines.

I suggest to extend this feature to all windows such as wxDialog, wxPanel, wxFrame, etc...

Thank you

**Comment by Randalphwa** [comment]
This portion of the issue has been implemented in PR #889

**Comment by Randalphwa** [comment]
See issue #892: Generate two-step construction for all forms · Issue #892 ·
KeyWorksRW/wxUiEditor (github.com)
<https://github.com/KeyWorksRW/wxUiEditor/issues/892>

I'm quite torn about the best way to handle this, and so while wxFrame
construction now does generate 2-step construction in the source code, it
does not provide a mechanism for the dev to insert code between the class
constructor and the call to Create(). As noted in the issue notes, enabling
that runs the risk of the dev creating code that will crash the program. I
do have it flagged for the next release which should be a much shorter time
frame now that I'm not trying to add an entire new language to generate.

**Reply by rossanoparis** [reply]
Kind @Randalphwa 

I'd say, if I were on you, I'd procede like wxDialog.
I'd only add an option for the two-step creation, no more than this.
Which is referred to what we normal call "forms" (wxDialog, wxPanel and wxFrames)
If the user whants to insert its own code, it must be inserted in its derived class, not in the main one created by your RAD tool.
For what it worth the absence of this option, which permits the two-step creation for al forms, is the main reason I can't use your RAD tool.

**Comment by Randalphwa** [comment]
I didn't realize this was a blocking issue for you. Sorry about that. I've
added the issue to the 1.1.1 milestone, and added some additional comments
about possible implementation method.

**Reply by rossanoparis** [reply]
I'm sorry, I didn't want to put pressure at all, that wasn't my intention, please don't get me wrong ...
I apologize for this misundertanding

I've wanted only to point the attention to a concept, which is "The developer shuldn't adapt its way to code because of the RAD tool".
Indeed the "two-step" creation can't be always avoided and it doesn't depend on the type of form used, and it is up to the developer whether to use it or not.

Regards
Rossano

**Comment by Randalphwa** [comment]
Ironically, I ran into the issue myself a couple of days ago when I needed
to create a wxFrame window that *required* 2-step construction and realized
that I couldn't use *wxUiEditor *to create it. So even if you hadn't sent
the email reminding me of the issue, I still would have added it to the
next milestone. I'm planning on adding some additional forms (both wxWindow
based and some non-window forms), so this needs to be dealt with before the
wxWindow forms get added.
